### DISCUSSION TITLE
[mini.pick] How to paste from system clipboard in SSH

### DISCUSSION THREAD

**Discussion by dhananjaylatkar** [nvim-mini/mini.nvim#1263] [General]
I use my remote machine over SSH and `<C-r>+` does not paste from HOST's clipboard, it pastes from remote machines clipboard.

**Comment by echasnovski** [comment]
Is there a way you can paste what you want in regular buffer?

**Comment by dhananjaylatkar** [comment]
`<C-S-v>` works in regular buffer. But not in mini.pick window though

**Reply by echasnovski** [reply]
Does `<C-r>*` work? This is another register that may store clipboard text.

**Reply by dhananjaylatkar** [reply]
no. it doesn't work. Both `<C-r>+` and `<C-r>*` paste from remote machine's clipboard.

**Reply by echasnovski** [reply]
Then unless you can get this text inside some register, there is no way of supplying it in the 'mini.pick' with default configuration. This is mostly because it looks like the clipboard text is transferred into the remote machine *only* after `<C-S-v>`.

There are at least two courses of action here:
- If it is a rare occurrence, then I'd settle on 1) paste into the current buffer; 2) delete pasted text with `d`; 3) open picker and paste with `<C-r>"`.
- If it is a common occurrence, I'd look at either automating previous step or exploring whether overriding `vim.paste` similar to how it is done in ['mini.pick'](https://github.com/echasnovski/mini.nvim/blob/a535342b6105a684f16114f598d44318e6a1d20f/lua/mini/pick.lua#L705-L709) itself. It probably should be made prior to `require('mini.pick').setup()` call and may utilize [`set_picker_query()`](https://github.com/echasnovski/mini.nvim/blob/a535342b6105a684f16114f598d44318e6a1d20f/doc/mini-pick.txt#L1376).

**Reply by dhananjaylatkar** [reply]
This is very common use case for anyone using ssh.

Why is normal paste disabled when picker is active? Can you provide config option to not disable it?

Also, can you provide example code for 2nd method?

**Reply by echasnovski** [reply]
> Why is normal paste disabled when picker is active? Can you provide config option to not disable it?

Because 'mini.pick' doesn't work with "normal" way of inputting text, instead processing every key separately. Not disabling paste with `<C-S-v>` will just paste text in the items window, not in the query.

> Also, can you provide example code for 2nd method?

Not today. There are reference examples in links of the previous comment.

**Reply by dhananjaylatkar** [reply]
> There are reference examples in links of the previous comment.

Not able to understand how to utilize those APIs.

Anyway, I'll try to use `:Pick grep` instead of `:Pick grep_live` when searching for particular text. Let's see if I can live with this workaround.

**Comment by KaspervdHeijden** [comment]
This is indead a rather annoying thing to deal with. I guess this is because the search buffer is catching keystrokes via `vim.keymap.set()` method, to save from opening a dedicated window. I suspect @echasnovski  choose this route for simplicity. 

For now I will try the solution below, as I'm also dealing with this. The snippet might be an example of overriding the `vim.paste` as mentioned earlier. If this does not work I'm _forced_ to go back to telescope.

Please note, do this **after**  calling setup!

```lua
                local paste_orig = vim.paste
                ---@diagnostic disable-next-line: duplicate-set-field
                vim.paste = function (...)
                    if not MiniPick.is_picker_active() then
                        return paste_orig(...)
                    end

                    -- Encountered inconsistent register values after copying from host machine
                    -- Could add more, please check `Pick registers` on strategic times to inspect state
                    for _, reg in ipairs({ '+', '.', '*' }) do
                        local content = vim.fn.getreg(reg) or ''

                        if content ~= '' then
                            MiniPick.set_picker_query({ content })
                            return
                        end
                    end

                    -- Yes, I know about <C-r>, will suppress that message, but show \something\ for feedback reasons
                    vim.notify('No content to paste', vim.log.levels.WARN)
                end
```

**Reply by fnandogp** [reply]
Works like a charm, thanks!
